From 819ff50bec2c14d3a18b0b63ebdcc58fe9e8bb3c Mon Sep 17 00:00:00 2001
From: Benedikt Ritter <britter@apache.org>
Date: Sun, 3 Mar 2013 13:14:38 +0000
Subject: [PATCH] [BEANUTILS-411] - BeanUtilsBean.setProperty throws
 IllegalArgumentException if getter of nested property returns null. Thanks to
 Marcus Zander

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/beanutils/trunk@1452036 13f79535-47bb-0310-9956-ffa450edef68
---
 pom.xml                                       |  4 ++
 src/changes/changes.xml                       |  3 +
 .../commons/beanutils/BeanUtilsBean.java      |  3 +
 .../beanutils/bugs/Jira411TestCase.java       | 59 +++++++++++++++++++
 4 files changed, 69 insertions(+)
 create mode 100644 src/test/java/org/apache/commons/beanutils/bugs/Jira411TestCase.java

diff --git a/pom.xml b/pom.xml
index cf43f6117..6e0c38d3e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -144,6 +144,10 @@
       <name>Alex Crown</name>
       <email></email>
     </contributor>    
+    <contributor>
+      <name>Marcus Zander</name>
+      <email>marcus.z@gmx.de</email>
+    </contributor>    
   </contributors>
 
   <dependencies>
diff --git a/src/changes/changes.xml b/src/changes/changes.xml
index 175863824..68fdaf272 100644
--- a/src/changes/changes.xml
+++ b/src/changes/changes.xml
@@ -40,6 +40,9 @@ The <action> type attribute can be add,update,fix,remove.
   <body>
 
     <release version="1.8.4" date="in SVN" description="Bug fix for 1.8.3">
+      <action dev="britter" type="fix" issue="BEANUTILS-411" due-to="Marcus Zander">
+         BeanUtilsBean.setProperty throws IllegalArgumentException if getter of nested property returns null
+      </action>
       <action dev="britter" type="update" issue="BEANUTILS-429" >
          Delete trailing white spaces and white spaces on empty lines from all files
       </action>
diff --git a/src/main/java/org/apache/commons/beanutils/BeanUtilsBean.java b/src/main/java/org/apache/commons/beanutils/BeanUtilsBean.java
index b04fa7c4b..beee7feed 100644
--- a/src/main/java/org/apache/commons/beanutils/BeanUtilsBean.java
+++ b/src/main/java/org/apache/commons/beanutils/BeanUtilsBean.java
@@ -901,6 +901,9 @@ public void setProperty(Object bean, String name, Object value)
         while (resolver.hasNested(name)) {
             try {
                 target = getPropertyUtils().getProperty(target, resolver.next(name));
+                if (target == null) { // the value of a nested property is null
+                    return;
+                }
                 name = resolver.remove(name);
             } catch (NoSuchMethodException e) {
                 return; // Skip this property setter
diff --git a/src/test/java/org/apache/commons/beanutils/bugs/Jira411TestCase.java b/src/test/java/org/apache/commons/beanutils/bugs/Jira411TestCase.java
new file mode 100644
index 000000000..66afff1c5
--- /dev/null
+++ b/src/test/java/org/apache/commons/beanutils/bugs/Jira411TestCase.java
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.commons.beanutils.bugs;
+
+import junit.framework.TestCase;
+
+import org.apache.commons.beanutils.BeanUtilsBean;
+
+/**
+ * BeanUtilsBean.setProperty throws IllegalArgumentException if getter of nested
+ * property returns null
+ *
+ * <p />
+ * See https://issues.apache.org/jira/browse/BEANUTILS-411
+ * <p />
+ *
+ * @version $Revision$ $Date$
+ */
+public class Jira411TestCase extends TestCase {
+
+    private BeanUtilsBean instance;
+    private DummyBean testBean;
+
+    protected void setUp() throws Exception {
+        instance = new BeanUtilsBean();
+        testBean = new DummyBean();
+    }
+
+    public void testSetProperty() throws Exception {
+        instance.setProperty(testBean, "imgLink.x", "1");
+    }
+
+    public class DummyBean {
+
+        private String imgLink = null;
+
+        public String getImgLink() {
+            return imgLink;
+        }
+
+        public void setImgLink(String imgLink) {
+            this.imgLink = imgLink;
+        }
+    }
+}
